File: Examples/ObjcDemo/RixSDKPodDemo/RixEngineObjcDemo/MainVC.m
#import "MainVC.h"
#import <AppTrackingTransparency/AppTrackingTransparency.h>
#import <AdSupport/AdSupport.h>
#import <AlxAds/AlxAds-Swift.h>
#import "AlxMainVC.h"
#import "AdmobMainVC.h"
#import "MaxMainVC.h"
#import "TopOnMainVC.h"
#import "LevelPlayMainVC.h"

@interface MainMenuItem : NSObject
@property (nonatomic, copy) NSString *title;
@property (nonatomic, copy) NSString *subtitle;
@property (nonatomic, copy) UIViewController *(^makeVC)(void);
@end

@implementation MainMenuItem
@end

@interface PaddingLabel : UILabel
@end

@implementation PaddingLabel
- (CGSize)intrinsicContentSize {
    CGSize size = [super intrinsicContentSize];
    return CGSizeMake(size.width + 16, size.height + 4);
}
- (void)drawTextInRect:(CGRect)rect {
    [super drawTextInRect:CGRectInset(rect, 8, 2)];
}
@end

@interface MainMenuCardCell ()
@property (nonatomic, strong) UIView *cardView;
@property (nonatomic, strong) UILabel *titleLabel;
@property (nonatomic, strong) UILabel *subtitleLabel;
@property (nonatomic, strong) UIImageView *chevronImageView;
@end

@implementation MainMenuCardCell

+ (NSString *)reuseID { return @"MainMenuCardCell"; }

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier {
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        [self setupUI];
    }
    return self;
}

- (void)setHighlighted:(BOOL)highlighted animated:(BOOL)animated {
    [super setHighlighted:highlighted animated:animated];
    CGFloat scale = highlighted ? 0.98 : 1.0;
    CGFloat alpha = highlighted ? 0.78 : 1.0;
    [UIView animateWithDuration:animated ? 0.15 : 0 animations:^{
        self.cardView.transform = CGAffineTransformMakeScale(scale, scale);
        self.cardView.alpha = alpha;
    }];
}

- (void)configureWithTitle:(NSString *)title subtitle:(NSString *)subtitle {
    self.titleLabel.text = title;
    self.subtitleLabel.text = subtitle;
    self.chevronImageView.tintColor = [UIColor colorWithRed:0.78 green:0.79 blue:0.84 alpha:1];
    self.titleLabel.textColor = [UIColor colorWithRed:0.12 green:0.13 blue:0.20 alpha:1];
}

- (void)setupUI {
    self.selectionStyle = UITableViewCellSelectionStyleNone;
    self.backgroundColor = [UIColor clearColor];
    self.contentView.backgroundColor = [UIColor clearColor];

    self.cardView = [[UIView alloc] init];
    self.cardView.translatesAutoresizingMaskIntoConstraints = NO;
    self.cardView.backgroundColor = [UIColor whiteColor];
    self.cardView.layer.cornerRadius = 18;
    self.cardView.layer.shadowColor = [UIColor blackColor].CGColor;
    self.cardView.layer.shadowOpacity = 0.02;
    self.cardView.layer.shadowOffset = CGSizeMake(0, 2);
    self.cardView.layer.shadowRadius = 8;

    self.titleLabel = [[UILabel alloc] init];
    self.titleLabel.translatesAutoresizingMaskIntoConstraints = NO;
    self.titleLabel.font = [UIFont systemFontOfSize:18 weight:UIFontWeightSemibold];
    self.titleLabel.textColor = [UIColor colorWithRed:0.12 green:0.13 blue:0.20 alpha:1];
    self.titleLabel.numberOfLines = 1;

    self.subtitleLabel = [[UILabel alloc] init];
    self.subtitleLabel.translatesAutoresizingMaskIntoConstraints = NO;
    self.subtitleLabel.font = [UIFont systemFontOfSize:13 weight:UIFontWeightRegular];
    self.subtitleLabel.textColor = [UIColor colorWithRed:0.56 green:0.57 blue:0.64 alpha:1];
    self.subtitleLabel.numberOfLines = 2;

    self.chevronImageView = [[UIImageView alloc] init];
    self.chevronImageView.translatesAutoresizingMaskIntoConstraints = NO;
    if (@available(iOS 13.0, *)) {
        self.chevronImageView.image = [UIImage systemImageNamed:@"chevron.right"];
    }
    self.chevronImageView.contentMode = UIViewContentModeScaleAspectFit;
    self.chevronImageView.tintColor = [UIColor colorWithRed:0.78 green:0.79 blue:0.84 alpha:1];

    [self.contentView addSubview:self.cardView];
    [self.cardView addSubview:self.titleLabel];
    [self.cardView addSubview:self.subtitleLabel];
    [self.cardView addSubview:self.chevronImageView];

    NSLayoutConstraint *cardBottom = [self.cardView.bottomAnchor constraintEqualToAnchor:self.contentView.bottomAnchor constant:-6];
    cardBottom.priority = UILayoutPriorityDefaultHigh;
    NSLayoutConstraint *subtitleBottom = [self.subtitleLabel.bottomAnchor constraintEqualToAnchor:self.cardView.bottomAnchor constant:-14];
    subtitleBottom.priority = UILayoutPriorityDefaultHigh;

    [NSLayoutConstraint activateConstraints:@[
        [self.cardView.leadingAnchor constraintEqualToAnchor:self.contentView.leadingAnchor constant:20],
        [self.cardView.trailingAnchor constraintEqualToAnchor:self.contentView.trailingAnchor constant:-20],
        [self.cardView.topAnchor constraintEqualToAnchor:self.contentView.topAnchor constant:6],
        cardBottom,

        [self.titleLabel.leadingAnchor constraintEqualToAnchor:self.cardView.leadingAnchor constant:16],
        [self.titleLabel.topAnchor constraintEqualToAnchor:self.cardView.topAnchor constant:14],
        [self.titleLabel.trailingAnchor constraintLessThanOrEqualToAnchor:self.chevronImageView.leadingAnchor constant:-12],

        [self.subtitleLabel.leadingAnchor constraintEqualToAnchor:self.titleLabel.leadingAnchor],
        [self.subtitleLabel.trailingAnchor constraintLessThanOrEqualToAnchor:self.chevronImageView.leadingAnchor constant:-12],
        [self.subtitleLabel.topAnchor constraintEqualToAnchor:self.titleLabel.bottomAnchor constant:6],
        subtitleBottom,

        [self.chevronImageView.trailingAnchor constraintEqualToAnchor:self.cardView.trailingAnchor constant:-16],
        [self.chevronImageView.centerYAnchor constraintEqualToAnchor:self.cardView.centerYAnchor],
        [self.chevronImageView.widthAnchor constraintEqualToConstant:14],
        [self.chevronImageView.heightAnchor constraintEqualToConstant:14]
    ]];
}

@end

@interface MainVC ()
@property (nonatomic, strong) NSArray<MainMenuItem *> *items;
@property (nonatomic, assign) CGFloat lastHeaderWidth;
@end

@implementation MainVC

- (void)viewDidLoad {
    [super viewDidLoad];
    self.navigationItem.title = nil;
    [self configureMainListAppearance];
    [self setupTopHeaderView];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self.navigationController setNavigationBarHidden:YES animated:animated];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    [self.navigationController setNavigationBarHidden:NO animated:animated];
}

- (void)viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];
    CGFloat currentWidth = self.tableView.bounds.size.width;
    if (currentWidth > 0 && fabs(currentWidth - self.lastHeaderWidth) > 0.5) {
        self.lastHeaderWidth = currentWidth;
        [self setupTopHeaderView];
    }
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(2.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [self requestATTPermission];
    });
}

- (NSArray<MainMenuItem *> *)items {
    if (!_items) {
        MainMenuItem *alx = [[MainMenuItem alloc] init];
        alx.title = NSLocalizedString(@"Alx_ad", @"");
        alx.subtitle = @"Direct integration with RixEngine ad serving.";
        alx.makeVC = ^UIViewController *{ return [[AlxMainVC alloc] init]; };

        MainMenuItem *admob = [[MainMenuItem alloc] init];
        admob.title = NSLocalizedString(@"admob_ad", @"");
        admob.subtitle = @"Google AdMob mediation integration.";
        admob.makeVC = ^UIViewController *{ return [[AdmobMainVC alloc] init]; };

        MainMenuItem *max = [[MainMenuItem alloc] init];
        max.title = NSLocalizedString(@"max_ad", @"");
        max.subtitle = @"AppLovin MAX mediation integration.";
        max.makeVC = ^UIViewController *{ return [[MaxMainVC alloc] init]; };

        MainMenuItem *topOn = [[MainMenuItem alloc] init];
        topOn.title = NSLocalizedString(@"topOn_ad", @"");
        topOn.subtitle = @"TopOn mediation integration.";
        topOn.makeVC = ^UIViewController *{ return [[TopOnMainVC alloc] init]; };

        MainMenuItem *levelPlay = [[MainMenuItem alloc] init];
        levelPlay.title = NSLocalizedString(@"levelPlay_ad", @"");
        levelPlay.subtitle = @"ironSource LevelPlay mediation integration.";
        levelPlay.makeVC = ^UIViewController *{ return [[LevelPlayMainVC alloc] init]; };

        _items = @[alx, admob, max, topOn, levelPlay];
    }
    return _items;
}

- (void)configureMainListAppearance {
    self.view.backgroundColor = [UIColor colorWithRed:0.95 green:0.95 blue:0.98 alpha:1.0];
    self.tableView.backgroundColor = self.view.backgroundColor;
    self.tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
    self.tableView.showsVerticalScrollIndicator = NO;
    self.tableView.contentInset = UIEdgeInsetsMake(6, 0, 24, 0);
    self.tableView.estimatedRowHeight = 90;
    self.tableView.rowHeight = UITableViewAutomaticDimension;
    [self.tableView registerClass:[MainMenuCardCell class] forCellReuseIdentifier:[MainMenuCardCell reuseID]];
    if (@available(iOS 15.0, *)) {
        self.tableView.sectionHeaderTopPadding = 0;
    }
}

- (void)setupTopHeaderView {
    CGFloat horizontalInset = 24;
    UIView *container = [[UIView alloc] init];
    container.backgroundColor = [UIColor clearColor];

    UIView *iconWrap = [[UIView alloc] init];
    iconWrap.backgroundColor = [UIColor colorWithRed:0.90 green:0.92 blue:0.98 alpha:1];
    iconWrap.layer.cornerRadius = 22;
    iconWrap.translatesAutoresizingMaskIntoConstraints = NO;

    UIImageView *iconImage = [[UIImageView alloc] init];
    iconImage.translatesAutoresizingMaskIntoConstraints = NO;
    iconImage.contentMode = UIViewContentModeScaleAspectFill;
    iconImage.layer.cornerRadius = 18;
    iconImage.clipsToBounds = YES;
    iconImage.image = [UIImage imageNamed:@"appLogo"];
    [iconWrap addSubview:iconImage];

    UILabel *titleLabel = [[UILabel alloc] init];
    titleLabel.translatesAutoresizingMaskIntoConstraints = NO;
    titleLabel.text = @"SDK Demo";
    titleLabel.font = [UIFont systemFontOfSize:17 weight:UIFontWeightBold];
    titleLabel.textColor = [UIColor colorWithRed:0.10 green:0.11 blue:0.20 alpha:1];

    // Version badge
    PaddingLabel *versionLabel = [[PaddingLabel alloc] init];
    versionLabel.translatesAutoresizingMaskIntoConstraints = NO;
    versionLabel.text = [self appVersionText];
    versionLabel.font = [UIFont systemFontOfSize:11 weight:UIFontWeightMedium];
    versionLabel.textColor = [UIColor colorWithRed:0.35 green:0.35 blue:0.55 alpha:1];
    versionLabel.backgroundColor = [UIColor colorWithRed:0.92 green:0.92 blue:0.97 alpha:1];
    versionLabel.layer.cornerRadius = 10;
    versionLabel.clipsToBounds = YES;
    
    // ObjC language badge
    UIView *ocBadge = [[UIView alloc] init];
    ocBadge.backgroundColor = [UIColor colorWithRed:0.93 green:0.96 blue:0.98 alpha:1];
    ocBadge.layer.cornerRadius = 10;
    ocBadge.clipsToBounds = YES;
    ocBadge.translatesAutoresizingMaskIntoConstraints = NO;

    UIImageView *objcIcon = [[UIImageView alloc] init];
    if (@available(iOS 13.0, *)) {
        UIImageSymbolConfiguration *config = [UIImageSymbolConfiguration configurationWithPointSize:11 weight:UIFontWeightMedium];
        objcIcon.image = [UIImage systemImageNamed:@"chevron.left.forwardslash.chevron.right" withConfiguration:config];
    }
    objcIcon.translatesAutoresizingMaskIntoConstraints = NO;
    objcIcon.tintColor = [UIColor colorWithRed:0.2 green:0.5 blue:0.8 alpha:1];
    objcIcon.contentMode = UIViewContentModeScaleAspectFit;

    UILabel *objcLabel = [[UILabel alloc] init];
    objcLabel.translatesAutoresizingMaskIntoConstraints = NO;
    objcLabel.text = @"ObjC";
    objcLabel.font = [UIFont systemFontOfSize:11 weight:UIFontWeightMedium];
    objcLabel.textColor = [UIColor colorWithRed:0.5 green:0.35 blue:0.55 alpha:1];

    UILabel *idfaLabel = [[UILabel alloc] init];
    idfaLabel.translatesAutoresizingMaskIntoConstraints = NO;
    idfaLabel.text = [NSString stringWithFormat:@"IDFA:%@", [[ASIdentifierManager sharedManager] advertisingIdentifier].UUIDString];
    idfaLabel.font = [UIFont systemFontOfSize:12 weight:UIFontWeightRegular];
    idfaLabel.textColor = [UIColor colorWithRed:0.64 green:0.66 blue:0.74 alpha:1];
    idfaLabel.numberOfLines = 2;

    UIStackView *titleStack = [[UIStackView alloc] initWithArrangedSubviews:@[titleLabel, versionLabel, objcIcon, objcLabel]];
    titleStack.translatesAutoresizingMaskIntoConstraints = NO;
    titleStack.axis = UILayoutConstraintAxisHorizontal;
    titleStack.alignment = UIStackViewAlignmentCenter;
    titleStack.spacing = 8;
    [titleStack setCustomSpacing:12 afterView:titleLabel];
    [titleStack setCustomSpacing:12 afterView:versionLabel];
    [titleStack setCustomSpacing:3 afterView:objcIcon];

    [container addSubview:iconWrap];
    [container addSubview:titleStack];
    [container addSubview:idfaLabel];

    NSLayoutConstraint *idfaTopPreferred = [idfaLabel.topAnchor constraintEqualToAnchor:iconWrap.bottomAnchor constant:14];
    idfaTopPreferred.priority = UILayoutPriorityDefaultHigh;
    NSLayoutConstraint *idfaBottomPreferred = [idfaLabel.bottomAnchor constraintEqualToAnchor:container.bottomAnchor constant:-8];
    idfaBottomPreferred.priority = UILayoutPriorityDefaultHigh;

    [NSLayoutConstraint activateConstraints:@[
        [iconWrap.leadingAnchor constraintEqualToAnchor:container.leadingAnchor constant:horizontalInset],
        [iconWrap.topAnchor constraintEqualToAnchor:container.topAnchor constant:6],
        [iconWrap.widthAnchor constraintEqualToConstant:44],
        [iconWrap.heightAnchor constraintEqualToConstant:44],

        [iconImage.centerXAnchor constraintEqualToAnchor:iconWrap.centerXAnchor],
        [iconImage.centerYAnchor constraintEqualToAnchor:iconWrap.centerYAnchor],
        [iconImage.widthAnchor constraintEqualToConstant:36],
        [iconImage.heightAnchor constraintEqualToConstant:36],

        [titleStack.leadingAnchor constraintEqualToAnchor:iconWrap.trailingAnchor constant:10],
        [titleStack.centerYAnchor constraintEqualToAnchor:iconWrap.centerYAnchor],
        [titleStack.trailingAnchor constraintLessThanOrEqualToAnchor:container.trailingAnchor constant:-horizontalInset],

        [objcIcon.widthAnchor constraintEqualToConstant:14],
        [objcIcon.heightAnchor constraintEqualToConstant:14],

        [idfaLabel.leadingAnchor constraintEqualToAnchor:container.leadingAnchor constant:horizontalInset],
        [idfaLabel.trailingAnchor constraintEqualToAnchor:container.trailingAnchor constant:-horizontalInset],
        [idfaLabel.topAnchor constraintGreaterThanOrEqualToAnchor:iconWrap.bottomAnchor constant:14],
        idfaTopPreferred,
        [idfaLabel.bottomAnchor constraintLessThanOrEqualToAnchor:container.bottomAnchor constant:-8],
        idfaBottomPreferred
    ]];

    CGFloat fittingWidth = self.tableView.bounds.size.width;
    if (fittingWidth > 0) {
        CGFloat textWidth = MAX(0, fittingWidth - (horizontalInset * 2));
        CGSize idfaSize = [idfaLabel sizeThatFits:CGSizeMake(textWidth, CGFLOAT_MAX)];
        CGFloat idfaHeight = ceil(idfaSize.height);
        CGFloat headerHeight = 6 + 44 + 14 + idfaHeight + 8;
        container.frame = CGRectMake(0, 0, fittingWidth, headerHeight);
        self.tableView.tableHeaderView = container;
    }
}

- (NSString *)appVersionText {
    return [AlxSdk getSDKVersion];
}

- (void)requestATTPermission {
    if (@available(iOS 14, *)) {
        [ATTrackingManager requestTrackingAuthorizationWithCompletionHandler:^(ATTrackingManagerAuthorizationStatus status) {
            [[NSUserDefaults standardUserDefaults] setBool:YES forKey:@"hasRequestedTrackingAuthorization"];
            NSString *idfa = [[ASIdentifierManager sharedManager] advertisingIdentifier].UUIDString;
            NSLog(@"idfa: %@", idfa);
        }];
    } else {
        NSString *idfa = [[ASIdentifierManager sharedManager] advertisingIdentifier].UUIDString;
        NSLog(@"idfa: %@", idfa);
    }
}

#pragma mark - UITableViewDataSource / UITableViewDelegate

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.items.count;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return UITableViewAutomaticDimension;
}

- (CGFloat)tableView:(UITableView *)tableView estimatedHeightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 88;
}

- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section {
    UIView *container = [[UIView alloc] init];
    container.backgroundColor = [UIColor clearColor];

    UILabel *label = [[UILabel alloc] init];
    label.translatesAutoresizingMaskIntoConstraints = NO;
    label.text = @"Ad Platforms";
    label.font = [UIFont systemFontOfSize:13 weight:UIFontWeightSemibold];
    label.textColor = [UIColor colorWithRed:0.46 green:0.47 blue:0.55 alpha:1];
    [container addSubview:label];

    [NSLayoutConstraint activateConstraints:@[
        [label.leadingAnchor constraintEqualToAnchor:container.leadingAnchor constant:24],
        [label.trailingAnchor constraintLessThanOrEqualToAnchor:container.trailingAnchor constant:-24],
        [label.bottomAnchor constraintLessThanOrEqualToAnchor:container.bottomAnchor constant:-8],
        [label.topAnchor constraintGreaterThanOrEqualToAnchor:container.topAnchor constant:6],
        [label.centerYAnchor constraintEqualToAnchor:container.centerYAnchor]
    ]];
    return container;
}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section {
    return 34;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    MainMenuCardCell *cell = [tableView dequeueReusableCellWithIdentifier:[MainMenuCardCell reuseID] forIndexPath:indexPath];
    MainMenuItem *item = self.items[indexPath.row];
    [cell configureWithTitle:item.title subtitle:item.subtitle];
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    MainMenuItem *item = self.items[indexPath.row];
    [self.navigationController pushViewController:item.makeVC() animated:YES];
}

@end

File: Examples/SwiftDemo/RixSDKSPMDemo/AlxAdsDemo/UI/MainVC.swift
//
//  MainVC.swift
//  AlxDemo
//
//  Created by liu weile on 2025/3/31.
//

import UIKit
import AppTrackingTransparency
import AdSupport
import AlxAds

struct MenuItem {
    let title: String
    let description: String
    let viewControllerType: UIViewController.Type
}

struct MenuSection {
    let title: String
    let items: [MenuItem]
}

// MARK: - MainVC

class MainVC: BaseUIViewController {

    private let menuSections: [MenuSection] = [
        MenuSection(
            title: "Ad Platforms",
            items: [
                MenuItem(title: "Alx Ad",
                         description: "Direct integration with RixEngine ad serving.",
                         viewControllerType: AlxMainVC.self),
                MenuItem(title: "Admob Ad",
                         description: "Google AdMob mediation integration.",
                         viewControllerType: AdmobMainVC.self),
                MenuItem(title: "Max Ad",
                         description: "AppLovin MAX mediation integration.",
                         viewControllerType: MaxMainVC.self),
//                MenuItem(title: "TopOn Ad",
//                         description: "TopOn mediation integration.",
//                         viewControllerType: TopOnMainVC.self),
                MenuItem(title: "LevelPlay Ad",
                         description: "IronSource LevelPlay mediation integration.",
                         viewControllerType: LevelPlayMainVC.self)
            ]
        ),
        MenuSection(
            title: "Internal",
            items: [
                MenuItem(title: "Test",
                         description: "Internal testing tools for QA validation.",
                         viewControllerType: TestMainVC.self)
            ]
        )
    ]

    private lazy var tableView: UITableView = {
        let t = UITableView(frame: .zero, style: .grouped)
        t.translatesAutoresizingMaskIntoConstraints = false
        t.delegate = self
        t.dataSource = self
        t.separatorStyle = .none
        t.backgroundColor = .clear
        t.showsVerticalScrollIndicator = false
        t.register(MenuCardCell.self, forCellReuseIdentifier: MenuCardCell.reuseID)
        t.sectionFooterHeight = 0
        t.contentInset = UIEdgeInsets(top: 0, left: 0, bottom: 20, right: 0)
        return t
    }()

    private let gradientLayer = CAGradientLayer()
    private var sectionHeaderCache: [Int: UIView] = [:]
    private lazy var cachedSDKVersion: String = AlxSdk.getSDKVersion()
    private lazy var cachedIdfaText: String = "IDFA:\(ASIdentifierManager.shared().advertisingIdentifier.uuidString)"

    // MARK: - Lifecycle

    override func viewDidLoad() {
        super.viewDidLoad()
        navigationController?.setNavigationBarHidden(true, animated: false)
        setupBackground()
        setupHeader()
        setupTableView()
        let appidString = AlxSdk.getAppID()
    }

    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        gradientLayer.frame = view.bounds
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        DemoMenuNavigation.warmupPlatformViewControllersIfNeeded()
        DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(2)) {
            self.requestATTPermission()
        }
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        navigationController?.setNavigationBarHidden(true, animated: animated)
    }

    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        navigationController?.setNavigationBarHidden(false, animated: animated)
    }

    // MARK: - Setup

    private func setupBackground() {
        gradientLayer.colors = [
            UIColor(red: 0.93, green: 0.94, blue: 1.0, alpha: 1).cgColor,
            UIColor(red: 0.96, green: 0.96, blue: 1.0, alpha: 1).cgColor,
            UIColor.white.cgColor
        ]
        gradientLayer.locations = [0, 0.35, 1]
        gradientLayer.startPoint = CGPoint(x: 0.5, y: 0)
        gradientLayer.endPoint = CGPoint(x: 0.5, y: 1)
        view.layer.insertSublayer(gradientLayer, at: 0)
    }

    private var headerBottomAnchor: NSLayoutYAxisAnchor!

    private func setupHeader() {
        let headerContainer = UIView()
        headerContainer.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(headerContainer)

        // Logo
        let logoView = UIImageView()
        logoView.image = UIImage(named: "Logo")
        logoView.contentMode = .scaleAspectFill
        logoView.clipsToBounds = true
        logoView.layer.cornerRadius = 22
        logoView.translatesAutoresizingMaskIntoConstraints = false

        // Title
        let titleLabel = UILabel()
        titleLabel.text = "SDK Demo"
        titleLabel.font = UIFont.systemFont(ofSize: 22, weight: .bold)
        titleLabel.textColor = UIColor(red: 0.1, green: 0.1, blue: 0.2, alpha: 1)
        titleLabel.translatesAutoresizingMaskIntoConstraints = false

        // Version badge
        let versionBadge = PaddingLabel()
        versionBadge.text = cachedSDKVersion
        versionBadge.font = UIFont.monospacedSystemFont(ofSize: 11, weight: .medium)
        versionBadge.textColor = UIColor(red: 0.35, green: 0.35, blue: 0.55, alpha: 1)
        versionBadge.backgroundColor = UIColor(red: 0.92, green: 0.92, blue: 0.97, alpha: 1)
        versionBadge.layer.cornerRadius = 10
        versionBadge.clipsToBounds = true
        versionBadge.translatesAutoresizingMaskIntoConstraints = false

        // Swift language badge
        let swiftBadge = UIView()
        swiftBadge.backgroundColor = UIColor(red: 0.96, green: 0.93, blue: 0.98, alpha: 1)
        swiftBadge.layer.cornerRadius = 10
        swiftBadge.clipsToBounds = true
        swiftBadge.translatesAutoresizingMaskIntoConstraints = false

        let swiftIcon = UIImageView()
        let swiftConfig = UIImage.SymbolConfiguration(pointSize: 11, weight: .medium)
        swiftIcon.image = UIImage(systemName: "swift", withConfiguration: swiftConfig)
        swiftIcon.tintColor = UIColor(red: 0.95, green: 0.35, blue: 0.2, alpha: 1)
        swiftIcon.translatesAutoresizingMaskIntoConstraints = false

        let swiftLabel = UILabel()
        swiftLabel.text = "Swift"
        swiftLabel.font = UIFont.systemFont(ofSize: 11, weight: .medium)
        swiftLabel.textColor = UIColor(red: 0.5, green: 0.35, blue: 0.55, alpha: 1)
        swiftLabel.translatesAutoresizingMaskIntoConstraints = false

        swiftBadge.addSubview(swiftIcon)
        swiftBadge.addSubview(swiftLabel)

        NSLayoutConstraint.activate([
            swiftIcon.leadingAnchor.constraint(equalTo: swiftBadge.leadingAnchor, constant: 6),
            swiftIcon.centerYAnchor.constraint(equalTo: swiftBadge.centerYAnchor),
            swiftLabel.leadingAnchor.constraint(equalTo: swiftIcon.trailingAnchor, constant: 3),
            swiftLabel.centerYAnchor.constraint(equalTo: swiftBadge.centerYAnchor),
            swiftLabel.trailingAnchor.constraint(equalTo: swiftBadge.trailingAnchor, constant: -6)
        ])

        // IDFA label
        let idfaLabel = UILabel()
        idfaLabel.text = cachedIdfaText
        idfaLabel.font = UIFont.monospacedSystemFont(ofSize: 11, weight: .regular)
        idfaLabel.textColor = UIColor(red: 0.5, green: 0.5, blue: 0.6, alpha: 1)
        idfaLabel.translatesAutoresizingMaskIntoConstraints = false

        // Decorative element (top-right blurred circle)
        let decoView = UIView()
        decoView.backgroundColor = UIColor(red: 0.7, green: 0.75, blue: 1.0, alpha: 0.15)
        decoView.layer.cornerRadius = 40
        decoView.translatesAutoresizingMaskIntoConstraints = false

        headerContainer.addSubview(decoView)
        headerContainer.addSubview(logoView)
        headerContainer.addSubview(titleLabel)
        headerContainer.addSubview(versionBadge)
        headerContainer.addSubview(swiftBadge)
        headerContainer.addSubview(idfaLabel)

        NSLayoutConstraint.activate([
            headerContainer.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
            headerContainer.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            headerContainer.trailingAnchor.constraint(equalTo: view.trailingAnchor),

            logoView.topAnchor.constraint(equalTo: headerContainer.topAnchor, constant: 16),
            logoView.leadingAnchor.constraint(equalTo: headerContainer.leadingAnchor, constant: 24),
            logoView.widthAnchor.constraint(equalToConstant: 44),
            logoView.heightAnchor.constraint(equalToConstant: 44),

            titleLabel.centerYAnchor.constraint(equalTo: logoView.centerYAnchor, constant: -1),
            titleLabel.leadingAnchor.constraint(equalTo: logoView.trailingAnchor, constant: 12),

            versionBadge.centerYAnchor.constraint(equalTo: titleLabel.centerYAnchor),
            versionBadge.leadingAnchor.constraint(equalTo: titleLabel.trailingAnchor, constant: 8),
            versionBadge.heightAnchor.constraint(equalToConstant: 20),

            swiftBadge.centerYAnchor.constraint(equalTo: titleLabel.centerYAnchor),
            swiftBadge.leadingAnchor.constraint(equalTo: versionBadge.trailingAnchor, constant: 6),
            swiftBadge.heightAnchor.constraint(equalToConstant: 20),

            idfaLabel.topAnchor.constraint(equalTo: logoView.bottomAnchor, constant: 10),
            idfaLabel.leadingAnchor.constraint(equalTo: headerContainer.leadingAnchor, constant: 24),
            idfaLabel.trailingAnchor.constraint(lessThanOrEqualTo: headerContainer.trailingAnchor, constant: -24),
            idfaLabel.bottomAnchor.constraint(equalTo: headerContainer.bottomAnchor, constant: -8),

            decoView.topAnchor.constraint(equalTo: headerContainer.topAnchor, constant: -10),
            decoView.trailingAnchor.constraint(equalTo: headerContainer.trailingAnchor, constant: 10),
            decoView.widthAnchor.constraint(equalToConstant: 80),
            decoView.heightAnchor.constraint(equalToConstant: 80)
        ])

        headerBottomAnchor = headerContainer.bottomAnchor
    }

    private func setupTableView() {
        view.addSubview(tableView)
        NSLayoutConstraint.activate([
            tableView.topAnchor.constraint(equalTo: headerBottomAnchor, constant: 4),
            tableView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            tableView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
            tableView.bottomAnchor.constraint(equalTo: view.bottomAnchor)
        ])
    }

    // MARK: - ATT

    func requestATTPermission() {
        if #available(iOS 14, *) {
            ATTrackingManager.requestTrackingAuthorization { status in
                UserDefaults.standard.set(true, forKey: "hasRequestedTrackingAuthorization")
                switch status {
                case .authorized:   print("ATT Authorized")
                case .denied:       print("ATT Denied")
                case .notDetermined:print("ATT Not Determined")
                case .restricted:   print("ATT Restricted")
                @unknown default:   print("ATT Unknown")
                }
                let idfa = ASIdentifierManager.shared().advertisingIdentifier.uuidString
                print("idfa:", idfa)
            }
        } else {
            let idfa = ASIdentifierManager.shared().advertisingIdentifier.uuidString
            print("idfa:", idfa)
        }
    }
}

// MARK: - UITableViewDataSource

extension MainVC: UITableViewDataSource {

    func numberOfSections(in tableView: UITableView) -> Int {
        return menuSections.count
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return menuSections[section].items.count
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard let cell = tableView.dequeueReusableCell(withIdentifier: MenuCardCell.reuseID, for: indexPath) as? MenuCardCell else {
            return UITableViewCell()
        }
        let item = menuSections[indexPath.section].items[indexPath.row]
        cell.configure(title: item.title, description: item.description)
        return cell
    }
}

// MARK: - UITableViewDelegate

extension MainVC: UITableViewDelegate {

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        let item = menuSections[indexPath.section].items[indexPath.row]
        DemoMenuNavigation.push(item.viewControllerType, from: navigationController)
    }

    func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        if let cached = sectionHeaderCache[section] {
            return cached
        }
        let container = UIView()
        let label = UILabel()
        label.text = menuSections[section].title
        label.font = UIFont.systemFont(ofSize: 13, weight: .semibold)
        label.textColor = UIColor(red: 0.5, green: 0.5, blue: 0.6, alpha: 1)
        label.translatesAutoresizingMaskIntoConstraints = false
        container.addSubview(label)
        NSLayoutConstraint.activate([
            label.leadingAnchor.constraint(equalTo: container.leadingAnchor, constant: 32),
            label.bottomAnchor.constraint(equalTo: container.bottomAnchor, constant: -4)
        ])
        sectionHeaderCache[section] = container
        return container
    }

    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return 36
    }

    func tableView(_ tableView: UITableView, heightForFooterInSection section: Int) -> CGFloat {
        return 4
    }

    func tableView(_ tableView: UITableView, viewForFooterInSection section: Int) -> UIView? {
        return UIView()
    }
}
